File: Week-1/Day-2/Problem7.java
// Given a string s, sort it in decreasing order based on the frequency of the characters. The frequency of a character is the number of times it appears in the string.

// Return the sorted string. If there are multiple answers, return any of them.

 

// Example 1:

// Input: s = "tree"
// Output: "eert"
// Explanation: 'e' appears twice while 'r' and 't' both appear once.
// So 'e' must appear before both 'r' and 't'. Therefore "eetr" is also a valid answer.
// Example 2:

// Input: s = "cccaaa"
// Output: "aaaccc"
// Explanation: Both 'c' and 'a' appear three times, so both "cccaaa" and "aaaccc" are valid answers.
// Note that "cacaca" is incorrect, as the same characters must be together.
// Example 3:

// Input: s = "Aabb"
// Output: "bbAa"
// Explanation: "bbaA" is also a valid answer, but "Aabb" is incorrect.
// Note that 'A' and 'a' are treated as two different characters.
 

// Constraints:

// 1 <= s.length <= 5 * 105
// s consists of uppercase and lowercase English letters and digits.

// https://leetcode.com/problems/sort-characters-by-frequency/description/

// Solution:
import java.util.*;


class Solution {
    public String frequencySort(String s) {
        int[] freq = new int[256];
        for(char c :s.toCharArray()){
            freq[c]++;
        }
        List<Character> chars = new ArrayList<>();
        for(char c = 0;c<256;c++){
            if(freq[c] > 0){
                chars.add(c);
            }
        }
         chars.sort((a, b) -> Integer.compare(freq[b], freq[a]));
         StringBuilder str = new StringBuilder();
         for(char c:chars){
            for(int i=0;i<freq[c];i++){
                str.append(c);
            }
         }
         return str.toString();
    }
}

File: Week-4/Day-2/Problem7.py
# Given a string s, partition s such that every 
# substring
#  of the partition is a 
# palindrome
# .

# Return the minimum cuts needed for a palindrome partitioning of s.

# Example 1:

# Input: s = "aab"
# Output: 1
# Explanation: The palindrome partitioning ["aa","b"] could be produced using 1 cut.
# Example 2:

# Input: s = "a"
# Output: 0
# Example 3:

# Input: s = "ab"
# Output: 1

# Constraints:

# 1 <= s.length <= 2000
# s consists of lowercase English letters only.


# https://leetcode.com/problems/palindrome-partitioning-ii/


class Solution:
    def minCut(self, s: str) -> int:
        n = len(s)
        dp = [0] * n
        isPalindrome = [[False] * n for _ in range(n)]

        for end in range(n):
            minCuts = end
            for start in range(end + 1):
                if s[start] == s[end] and (end - start <= 2 or isPalindrome[start + 1][end - 1]):
                    isPalindrome[start][end] = True
                    minCuts = 0 if start == 0 else min(minCuts, dp[start - 1] + 1)
            dp[end] = minCuts

        return dp[n - 1]